### ISSUE TITLE
Make master green again!

### ISSUE THREAD

**Issue by lgierth** [ipfs/go-ipfs#2002]
There are a lot of very horrible CI failures lately, ranging from time-based, to plain worrying (like `ipfs add -r` hash mismatches).

- Let's look at each failure (!)
- Collect types of failure in this issue
- Fix them

### the failures

- [x] #2026 sharness t0020-init.sh fails to cleanup on OSX
- [ ] #2027 t0043 fails occasionally
- [ ] #2028 t0080-repo.sh fails on osx CI

**Comment by rht** [comment]
Some specimens can be found in #1989 (excluding the travis_wait commits).

There are 3 major types from this PR:
1. https://travis-ci.org/ipfs/go-ipfs/jobs/92502227 `ipfs add -r --hidden mountdir/planets >actual` stalled.
2. https://travis-ci.org/ipfs/go-ipfs/jobs/92499693 one of the auto-gc test fails.
3. ipfs refs -r counts 77 instead of 79.

I don't see more beyond the 3 types.
The rest is usual statistical goroutine errs.
All happens only in osx sharness.

**Comment by rht** [comment]
(#1984 has the most recent test I ran, and they all somehow passed)

**Comment by jbenet** [comment]
:+1: yes please

**Comment by lgierth** [comment]
#2027 t0043 fails occasionally

**Comment by cryptix** [comment]
I'm seeing this in t0060-daemon.sh quite often.

```
not ok 19 - 'ipfs daemon' should be able to run with a pipe attached to stdin (issue #861)
#
#         yes | ipfs daemon --init >stdin_daemon_out 2>stdin_daemon_err &
#         pollEndpoint -ep=/version -v -tout=1s -tries=10 >stdin_poll_apiout 2>stdin_poll_apierr &&
#         test_kill_repeat_10_sec $! ||
#         test_fsh cat stdin_daemon_out || test_fsh cat stdin_daemon_err || test_fsh cat stdin_poll_apiout || test_fsh cat stdin_poll_apierr
#
```

passing one in 5 times maybe, leaving a ipfs binary running n the system in failure case which is confusing on the next run. Digging deeper, the daemon simply didn't start up in time... (Also found small print bug in pollEndpoint the meantime, PR incoming)

**Comment by whyrusleeping** [comment]
@cryptix t0060 fails if you have a daemon already running.

**Comment by lgierth** [comment]
Doesn't properly kill the daemon after some test in https://circleci.com/gh/ipfs/go-ipfs/1981

```
expecting success: 
		IPFS_PID=$! &&
		pollEndpoint -ep=/version -host=$ADDR_API -v -tout=1s -tries=60 2>poll_apierr > poll_apiout ||
		test_fsh cat actual_daemon || test_fsh cat daemon_err || test_fsh cat poll_apierr || test_fsh cat poll_apiout
	
> cat actual_daemon
Initializing daemon...
Swarm listening on /ip4/10.0.3.235/tcp/13212
Swarm listening on /ip4/10.0.4.1/tcp/13212
Swarm listening on /ip4/127.0.0.1/tcp/13212
API server listening on /ip4/127.0.0.1/tcp/6312

> cat daemon_err
Error: serveHTTPGateway: manet.Listen(/ip4/127.0.0.1/tcp/9312) failed: listen tcp4 127.0.0.1:9312: bind: address already in use

> cat poll_apierr
00:33:29.862 DEBUG pollEndpoi: starting at %s, tries: %d, timeout: %s, url: %s2015-11-21 00:33:29.862385236 +0000 UTC 60 1s {http  <nil> 127.0.0.1:6312 /version   } main.go:57
00:33:29.862 DEBUG pollEndpoi: get failed: Get http://127.0.0.1:6312/version: dial tcp 127.0.0.1:6312: getsockopt: connection refused main.go:66
00:33:30.863 DEBUG pollEndpoi: get failed: Get http://127.0.0.1:6312/version: dial tcp 127.0.0.1:6312: getsockopt: connection refused main.go:66
[snip]

> cat poll_apiout

not ok 6 - 'ipfs daemon' is ready
```

**Comment by cryptix** [comment]
> @cryptix t0060 fails if you have a daemon already running.

Yes, but that failure looks different, the first cases fail.

**Comment by lgierth** [comment]
New failure on OSX: https://travis-ci.org/ipfs/go-ipfs/jobs/96669162#L147

```
--- FAIL: TestReconnect5 (1.94s)

	reconnect_test.go:231: host 0 <peer.ID bBGhBC> has 4 conns! not zero.

FAIL
```

And a timeout on OSX sharness: https://travis-ci.org/ipfs/go-ipfs/jobs/96669163

```
expecting success: 

		mkdir -p testdir &&

		echo "hello test" >testdir/test.txt &&

		ipfs add -r testdir &&

        curl -i "http://localhost:$PORT_API/api/v0/refs?arg=QmTcJAn3JP8ZMAKS6WS75q8sbTyojWKbxcUHgLYGWur4Ym&stream-channels=true&encoding=text" >actual_output

	

No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.
```

**Comment by jbenet** [comment]
I like xmas and all, but the red is getting so annoying :S

**Comment by rht** [comment]
Wouldn't this be most effective if debugged by someone on osx? Is the err specific to travis osx only?

**Comment by rht** [comment]
Saw a green on 0.3.8: https://github.com/ipfs/go-ipfs/pull/2081

**Comment by rht** [comment]
Green 3 times in a row in #2081

**Comment by rht** [comment]
0.3.9, 2 greens out of 10 so far
1. https://travis-ci.org/ipfs/go-ipfs/builds/97323648
2. https://travis-ci.org/ipfs/go-ipfs/builds/97324151
3. https://travis-ci.org/ipfs/go-ipfs/builds/97324526
4. https://travis-ci.org/ipfs/go-ipfs/builds/97325414
5. https://travis-ci.org/ipfs/go-ipfs/builds/97325761
6. https://travis-ci.org/ipfs/go-ipfs/builds/97325817
7. https://travis-ci.org/ipfs/go-ipfs/builds/97325907
8. https://travis-ci.org/ipfs/go-ipfs/builds/97325976
9. https://travis-ci.org/ipfs/go-ipfs/builds/97326107
10. https://travis-ci.org/ipfs/go-ipfs/builds/97326131

**Comment by chriscool** [comment]
Yeah, it should be possible to bisect this by doing many tests on each commit instead of just one

There is also this project that has a bisect that uses Bayesian Search Theory to find intermittent bugs:

https://github.com/Ealdwulf/bbchop

Yeah, it has not been updated for a long time but it might still be interesting.

**Comment by chriscool** [comment]
PR #2040 should have fixed:  #2026 sharness t0020-init.sh fails to cleanup on OSX

**Comment by rht** [comment]
@chriscool the bisection alg code you linked is only possible if it can be run on travis. Or if the test fails are reproducible on osx.

> PR #2040 should have fixed: #2026 sharness t0020-init.sh fails to cleanup on OSX

ok I will check 10x if this cures the test fails in general.

**Comment by chriscool** [comment]
> @chriscool the bisection alg code you linked is only possible if it can be run on travis. Or if the test fails are reproducible on osx.

Yeah, unfortunately it looks like CI services like Travis and Circle don't have a nice bisecting service (maybe we should suggest it to them), but I thought it could be interesting to talk about bbchop while at it.

>>    PR #2040 should have fixed: #2026 sharness t0020-init.sh fails to cleanup on OSX
>
> ok I will check 10x if this cures the test fails in general.

Great, thanks!

**Comment by rht** [comment]
"git bisect bad" at 0.3.10: https://github.com/ipfs/go-ipfs/pull/2084.
3 greens:
- https://travis-ci.org/ipfs/go-ipfs/builds/97333686
- https://travis-ci.org/ipfs/go-ipfs/builds/97335816
- https://travis-ci.org/ipfs/go-ipfs/builds/97336443

gc trigger errs:
- https://travis-ci.org/ipfs/go-ipfs/jobs/97335899
- https://travis-ci.org/ipfs/go-ipfs/jobs/97336100

refs -r errs:
- https://travis-ci.org/ipfs/go-ipfs/jobs/97336791

random-files errs:
- https://travis-ci.org/ipfs/go-ipfs/jobs/97336873

stalled:
- https://travis-ci.org/ipfs/go-ipfs/jobs/97335954
- https://travis-ci.org/ipfs/go-ipfs/jobs/97336261
- https://travis-ci.org/ipfs/go-ipfs/jobs/97336368

64 commits between 0.3.9 https://github.com/ipfs/go-ipfs/commit/220d6b168f06c6e6bdb458effa2a27e8f5d7c439 and 0.3.10 https://github.com/ipfs/go-ipfs/commit/f9dc4c726b770199f4ee64d97775d5fe8122814e.

**Comment by rht** [comment]
current master (https://github.com/ipfs/go-ipfs/commit/df65baefca0375e0602a6afd178d36dabecb8f65):
4 greens:
- https://travis-ci.org/ipfs/go-ipfs/builds/97335047
- https://travis-ci.org/ipfs/go-ipfs/builds/97335173
- https://travis-ci.org/ipfs/go-ipfs/builds/97335202
- https://travis-ci.org/ipfs/go-ipfs/builds/97335364

gc trigger errs:
- https://travis-ci.org/ipfs/go-ipfs/jobs/97335507
- https://travis-ci.org/ipfs/go-ipfs/builds/97334764

3 stalled:
- https://travis-ci.org/ipfs/go-ipfs/builds/97334674
- https://travis-ci.org/ipfs/go-ipfs/builds/97335414
- https://travis-ci.org/ipfs/go-ipfs/builds/97335443

linux sharness:
- https://travis-ci.org/ipfs/go-ipfs/builds/97334411

I bet the most likely culprit is the gc PR https://github.com/ipfs/go-ipfs/pull/1495.
Will check before-after this one first before doing the usual bisect.

**Comment by rht** [comment]
Actually, pre-gc has build errs as well

green
- https://travis-ci.org/ipfs/go-ipfs/builds/97369380
- https://travis-ci.org/ipfs/go-ipfs/builds/97369631
- https://travis-ci.org/ipfs/go-ipfs/builds/97369690
- https://travis-ci.org/ipfs/go-ipfs/builds/97369859
- https://travis-ci.org/ipfs/go-ipfs/builds/97369726

stalled:
- https://travis-ci.org/ipfs/go-ipfs/builds/97368603
- https://travis-ci.org/ipfs/go-ipfs/builds/97369482
- https://travis-ci.org/ipfs/go-ipfs/builds/97369652

refs -r:
- https://travis-ci.org/ipfs/go-ipfs/builds/97369432

add bigfile and cat:
- https://travis-ci.org/ipfs/go-ipfs/builds/97369671

There are 21 commits between https://github.com/ipfs/go-ipfs/commits/pregc and 0.3.9

**Comment by rht** [comment]
git bisect bad: https://github.com/ipfs/go-ipfs/commits/log-hang-closenotify, https://github.com/ipfs/go-ipfs/pull/2088

green
- https://travis-ci.org/ipfs/go-ipfs/builds/97376242
- https://travis-ci.org/ipfs/go-ipfs/builds/97378570
- https://travis-ci.org/ipfs/go-ipfs/builds/97378619
- https://travis-ci.org/ipfs/go-ipfs/builds/97378726
- https://travis-ci.org/ipfs/go-ipfs/builds/97378849
- https://travis-ci.org/ipfs/go-ipfs/builds/97378928
- https://travis-ci.org/ipfs/go-ipfs/builds/97379113

stalled
- https://travis-ci.org/ipfs/go-ipfs/builds/97378760
- https://travis-ci.org/ipfs/go-ipfs/builds/97378862
- https://travis-ci.org/ipfs/go-ipfs/builds/97379137

**Comment by rht** [comment]
I have narrowed down the osx build stall to this PR https://github.com/ipfs/go-ipfs/pull/1914.

**Comment by jbenet** [comment]
@rht _great_ detective work! but wait, one problem: 

the closenotify thing caused hangs in <go1.5.1. we moved to go1.5.2, and i updated the `travis.yml` to use 1.5.2 some time after #1914 (shoud've been in same PR but wasn't). this may cause false positives for #1914 -- try rerunning that one with go1.5.2 in this line: https://github.com/ipfs/go-ipfs/blob/master/.travis.yml#L10

**Comment by rht** [comment]
reran with 1.5.2, still build errs/stalls: https://github.com/ipfs/go-ipfs/pull/2088

**Comment by rht** [comment]
The gc errs are likely due to race condition between periodic gc and pinning, when added files are gc-ed before they are pinned.

**Comment by Kubuxu** [comment]
stalled: https://travis-ci.org/ipfs/go-ipfs/jobs/99128512
on t0045-ls.sh

**Comment by thelinuxkid** [comment]
I am having the timeout issue with sharness tests too #2119. #2120 adds fuse support for Travis CI but also switches the Ubuntu version to Trusty. However, I ran the tests multiple times (go1.5.2) and the timeout is my only issue, albeit, intermittent -- redoing a test can eventually yield green: https://travis-ci.org/thelinuxkid/go-ipfs/builds/98464151. https://github.com/thelinuxkid/travis-rerun might help with debugging. I was able to build the Trusty image Travis uses. I will try and see if I can replicate the timeout there when I am in town again next week.
